File: AppleMusicAPIcall/app/src/main/java/com/example/applemusicapicall/RetrofitCLient.java
package com.example.applemusicapicall;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.converter.gson.GsonConverterFactory;

//is the client (like our virtual web browser)
public class RetrofitCLient {

    public static final String BASE_URL = "https://itunes.apple.com/";
    private static retrofit2.Retrofit retrofit = null;

    private RetrofitCLient(){

    }

    public static retrofit2.Retrofit setRetrofit() {

        if(retrofit ==null){
            // LOG HTTP REQUEST & RESPONSE WITH LOGGING INTERCEPTOR

            HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
            // set your desired log level
            // //OkHttp’s logging interceptor has four log levels: NONE, BASIC, HEADERS, BODY

            //Level.BODY This is the most complete log level and will print out every related
            // information for your request and response.
            logging.setLevel(HttpLoggingInterceptor.Level.BODY);

            OkHttpClient httpClient = new OkHttpClient.Builder()
                    .addInterceptor(logging) // add logging as last interceptor
                    .build();
             retrofit= new retrofit2.Retrofit.Builder()
                    .baseUrl(BASE_URL)  //base url
                    .addConverterFactory(GsonConverterFactory.create()) //where we define our gson library to parse object
                     .client(httpClient)
                    .build();
        }
        return retrofit;
    }
}

File: WeatherApp2/app/src/main/java/com/example/weatherapp/retrofit_api/RetrofitClient.kt
package com.example.weatherapp.retrofit_api

import android.content.Context
import com.example.weatherapp.retrofit_api.checkNetworkConnection.NetworkConnectionInterceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitClient {
    @Volatile
    private var RETROFIT_INSTANCE: Retrofit? = null
    const val BASE_URL = "https://weather.exam.bottlerocketservices.com/"

   private fun getClient(mContext:Context) : Retrofit {
        return RETROFIT_INSTANCE ?: synchronized(this) {
            val interceptor = HttpLoggingInterceptor()
            interceptor.setLevel(HttpLoggingInterceptor.Level.BODY)

            val client = OkHttpClient.Builder()
                    .build()
            val retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build()
            RETROFIT_INSTANCE = retrofit
            retrofit
        }
    }

    //Retrofit service generator
    public fun<T> getWeatherService(service: Class<T>,mContext:Context): T {
        return getClient(mContext).create(service)
    }
}